kusto/test/etoe/ingestion_status_e2e_test.go

```diff
@@ -0,0 +1,395 @@
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/Azure/azure-kusto-go/kusto"
+	"github.com/Azure/azure-kusto-go/kusto/ingest"
+	"github.com/Azure/azure-kusto-go/kusto/unsafe"
+)
+
+const (
+	tableName string = "GolangStatusReportingTest"
+	scheme    string = " (rownumber:int, rowguid:string, xdouble:real, xfloat:real, xbool:bool, xint16:int, xint32:int, xint64:long, xuint8:long, xuint16:long, xuint32:long, xuint64:long, xdate:datetime, xsmalltext:string, xtext:string, xnumberAsText:string, xtime:timespan, xtextWithNulls:string, xdynamicWithNulls:dynamic)"
+	csvFile   string = "testdata/dataset.csv"
+	verbose   bool   = false
+)
+
+var (
+	initDone   bool = false
+	initFailed bool = true
+	testConf   Config
+)
+
+func initOnce() error {
+	if !initDone {
+		initDone = true
+
+		testConf, err := NewConfig()
+		if err != nil {
+			return fmt.Errorf("end to end tests disabled: missing config.json file in etoe directory or test environment not set - %s", err)
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		ctx, cancel = context.WithTimeout(ctx, 1*time.Minute)
+		defer cancel()
+
+		client, err := kusto.New(testConfig.Endpoint, testConfig.Authorizer)
+		if err != nil {
+			return err
+		}
+
+		// Drop the old table if exists
+		dropStmt := kusto.NewStmt(".drop table ", kusto.UnsafeStmt(unsafe.Stmt{Add: true})).UnsafeAdd(tableName).Add(" ifexists")
+		_, err = client.Mgmt(ctx, testConf.Database, dropStmt)
+		if err != nil {
+			return fmt.Errorf("failed to drop the old table:\n" + err.Error())
+		}
+
+		// Create a database
+		createStmt := kusto.NewStmt(".create table ", kusto.UnsafeStmt(unsafe.Stmt{Add: true})).UnsafeAdd(tableName).UnsafeAdd(scheme)
+		_, err = client.Mgmt(ctx, testConf.Database, createStmt)
+		if err != nil {
+			return fmt.Errorf("failed to create a table:\n" + err.Error())
+		}
+
+		// Change the ingetion batching time
+		batchingStmt := kusto.NewStmt(".alter table ", kusto.UnsafeStmt(unsafe.Stmt{Add: true})).UnsafeAdd(tableName).Add(" policy ingestionbatching @'{ \"MaximumBatchingTimeSpan\": \"00:00:25\", \"MaximumNumberOfItems\": 500, \"MaximumRawDataSizeMB\": 1024 }' ")
+		_, err = client.Mgmt(ctx, testConf.Database, batchingStmt)
+		if err != nil {
+			return fmt.Errorf("failed to reduce the default batching time\n" + err.Error())
+		}
+
+		initFailed = false
+	}
+
+	if initFailed {
+		return fmt.Errorf("Init once failed in the past")
+	}
+
+	return nil
+}
+
+func TestIgestionFromFileWithStatusReportingQueued(t *testing.T) {
+	err := initOnce()
+	if err != nil {
+		t.Skipf("Skipping tests: %s", err)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ctx, cancel = context.WithTimeout(ctx, 3*time.Minute)
+	defer cancel()
+
+	client, err := kusto.New(testConfig.Endpoint, testConfig.Authorizer)
+	if err != nil {
+		panic(err)
+	}
+
+	ingestor, err := ingest.New(client, testConfig.Database, tableName)
+	if err != nil {
+		panic(err)
+	}
+
+	count := 5

```

Comment:
You can do this in this manner, but it feels odd.  You almost never use arrays in Go.  And you usually only use C style loops for counting usually once for a set of activities.

This is more in the Go fashion.
```go
const numTests = 5

results := make([]*Result, 0, numTests)

for  i := 0; i < numTests, i++ {
  res, err := ingestor.FromFile(ctx, csvFile, ingest.ReportResultToTable())
  if err != nil {
    t.Fatalf("TestIgestionFromFileWithStatusReportingQueued: got err == %q, want err == nil", err)
  }
  results = append(results, res)
}

for i, res := range results {
  if err := res.Wait(ctx); err != nil {
    t.Errorf("TestIgestionFromFileWithStatusReportingQueued(ingestion %d): got err == %q, want err == nil", i, err)
  } 
}
```
This still executes in parallel, eliminates the channels array and errors array, documents the errors better, and has a more Go feel.

Reply:
I don't see what's the big difference with the exception of calling make.
Also, your example has an error since wait returns a channel not an error, so I still need the channel array or I loose parallelism.

Reply:
Fixed the bug, I think.  Same code, I just changed it to use a chan error.
The difference is: less loops, doesn't use arrays (which is uncommon) and a single slice instead of two arrays.

But not worth spending a bunch of time on, up to you.